src/main/java/org/tdl/vireo/model/repo/impl/ActionLogRepoImpl.java

```diff
@@ -1,76 +1,92 @@
+import org.tdl.vireo.model.ActionLog;
+import org.tdl.vireo.model.Submission;
+import org.tdl.vireo.model.User;
+import org.tdl.vireo.model.repo.ActionLogRepo;
+import org.tdl.vireo.model.repo.SubmissionRepo;
+import org.tdl.vireo.model.repo.custom.ActionLogRepoCustom;
+
+import edu.tamu.weaver.data.model.repo.impl.AbstractWeaverRepoImpl;
+import edu.tamu.weaver.response.ApiResponse;
+
+public class ActionLogRepoImpl extends AbstractWeaverRepoImpl<ActionLog, ActionLogRepo> implements ActionLogRepoCustom {
+
+    @Autowired
+    private ActionLogRepo actionLogRepo;
+
+    @Autowired
+    private SubmissionRepo submissionRepo;
+
+    @Autowired
+    private SimpMessagingTemplate simpMessagingTemplate;
+
+    @Override
+    public ActionLog create(Submission submission, User user, Calendar actionDate, String entry, boolean privateFlag) {
+        ActionLog log = actionLogRepo.save(new ActionLog(submission.getSubmissionStatus(), user, actionDate, entry, privateFlag));
+        submission.addActionLog(log);
+        submission.setLastAction(log);
+        submissionRepo.save(submission);
+        simpMessagingTemplate.convertAndSend("/channel/submission/" + submission.getId() + "/action-logs", new ApiResponse(SUCCESS, log));
+        return log;
+    }
+
+    @Override
+    public ActionLog create(Submission submission, Calendar actionDate, String entry, boolean privateFlag) {
+        ActionLog log = actionLogRepo.save(new ActionLog(submission.getSubmissionStatus(), actionDate, entry, privateFlag));
+        submission.addActionLog(log);
+        submission.setLastAction(log);
+        submissionRepo.save(submission);
+        simpMessagingTemplate.convertAndSend("/channel/submission/" + submission.getId() + "/action-logs", new ApiResponse(SUCCESS, log));
+        return log;
+    }
+
+    @Override
+    public ActionLog createPublicLog(Submission submission, User user, String entry) {
+        return create(submission, user, Calendar.getInstance(), entry, false);
+    }
+
+    @Override
+    public ActionLog createAdvisorPublicLog(Submission submission, String entry) {
+        return create(submission, Calendar.getInstance(), entry, false);
+    }
+
+    @Override
+    public ActionLog createPrivateLog(Submission submission, User user, String entry) {
+        return create(submission, user, Calendar.getInstance(), entry, true);
+    }
+
+    @Override
+    public void delete(ActionLog actionLog) {
+        Submission submission = submissionRepo.findByActionLogsId(actionLog.getId());
+        submission.removeActionLog(actionLog);
+
+        List<ActionLog> actionLogs = new ArrayList<>(submission.getActionLogs());
+
+        if (!actionLogs.isEmpty() && Objects.nonNull(submission.getLastAction())
+                && submission.getLastAction().getId().equals(actionLog.getId())) {
+            actionLogs.sort(Comparator.comparing(ActionLog::getActionDate));
+            submission.setLastAction(actionLogs.get(actionLogs.size() - 1));
```

Comment:
This is very expensive.
It would be better to have an SQL that fetches the last log rather than having to load all of the logs here and sort them here then get the last log.

Reply:
A query would be more efficient.

Reply:
Optimization here is beyond the scope of this PR. We are relying on JPA to fetch all action logs for a given submission in order to remove one and if the last set the previous as last action. This possible with a delete JPA named query and an update JPA named query which both sort desc by date.

Reply:
I did not have time to update. It does only occur when a new action log is added. If the number of action logs per submission grow to where adding a new action log takes an unacceptable amount of time, here is where to look.